ui/src/shared/components/OptIn.js

```diff
@@ -0,0 +1,182 @@
+import React, {Component, PropTypes} from 'react'
+import classnames from 'classnames'
+
+import uuid from 'node-uuid'
+
+import ClickOutsideInput from 'shared/components/ClickOutsideInput'
+
+class OptIn extends Component {
+  constructor(props) {
+    super(props)
+
+    const {customValue, fixedValue} = props
+
+    this.state = {
+      useCustomValue: customValue !== '',
+      fixedValue,
+      customValue,
+    }
+
+    this.id = uuid.v4()
+    this.isCustomValueInputFocused = false
+
+    this.useFixedValue = ::this.useFixedValue
+    this.useCustomValue = ::this.useCustomValue
+    this.considerResetCustomValue = ::this.considerResetCustomValue
+    this.setCustomValue = ::this.setCustomValue
+    this.setValue = ::this.setValue
+  }
+
+  useFixedValue() {
+    this.setState({useCustomValue: false, customValue: ''}, () =>
+      this.setValue()
+    )
+  }
+
+  useCustomValue() {
+    this.setState({useCustomValue: true}, () => this.setValue())
+  }
+
+  handleClickFixedValueField() {
+    return () => this.useFixedValue()
+  }
+
+  handleClickToggle() {
+    return () => {
```

Comment:
As in, get rid of the `return` here and just pass in a reference to this method in the `onClick` rather than invoking it? If that's what you mean, I did it this way to keep it consistent for all UI event handlers, but am not tied to it.